Sources/Filters/General/ShrinkPolyData/index.js

```diff
@@ -0,0 +1,338 @@
+      for (let i = 0; i < cellSizes.length; i++) {
+        numNewPts += (cellSizes[i] - 1) * 2;
+      }
+    }
+
+    if (inPolys) {
+      const cellSizes = inPolys.getCellSizes();
+      for (let i = 0; i < cellSizes.length; i++) {
+        numNewPts += cellSizes[i];
+      }
+    }
+
+    if (inStrips) {
+      const cellSizes = inStrips.getCellSizes();
+      for (let i = 0; i < cellSizes.length; i++) {
+        numNewPts += (cellSizes[i] - 2) * 3;
+      }
+    }
+
+    const newPointsData = new Float32Array(numNewPts * 3);
+    const newPoints = vtkPoints.newInstance();
+    newPoints.setData(newPointsData, 3);
+
+    const newVerts = vtkCellArray.newInstance();
+    const newLines = vtkCellArray.newInstance();
+    const newPolys = vtkCellArray.newInstance();
+
+    let outCount = 0;
+    const center = [0, 0, 0];
+
+    // Process vertices (no shrinking necessary)
+    if (inVerts) {
+      const vertData = inVerts.getData();
+      const newVertData = [];
+
+      for (let i = 0; i < vertData.length; ) {
+        const npts = vertData[i];
+        newVertData.push(npts);
+
+        for (let j = 1; j <= npts; j++) {
+          const ptId = vertData[i + j];
+          // Copy point
+          newPointsData[outCount * 3] = inPts[ptId * 3];
+          newPointsData[outCount * 3 + 1] = inPts[ptId * 3 + 1];
+          newPointsData[outCount * 3 + 2] = inPts[ptId * 3 + 2];
+
+          newVertData.push(outCount);
+          outCount++;
+        }
+        i += npts + 1;
+      }
+
+      newVerts.setData(new Uint32Array(newVertData));
+    }
+
+    // Process lines (shrink and split)
+    if (inLines) {
+      const lineData = inLines.getData();
+      const newLineData = [];
+
+      for (let i = 0; i < lineData.length; ) {
+        const npts = lineData[i];
+
+        for (let j = 0; j < npts - 1; j++) {
+          const ptId1 = lineData[i + j + 1];
+          const ptId2 = lineData[i + j + 2];
+
+          const p1 = [
+            inPts[ptId1 * 3],
+            inPts[ptId1 * 3 + 1],
+            inPts[ptId1 * 3 + 2],
+          ];
+          const p2 = [
+            inPts[ptId2 * 3],
+            inPts[ptId2 * 3 + 1],
+            inPts[ptId2 * 3 + 2],
+          ];
+
+          const [shrunkP1, shrunkP2] = publicAPI.shrinkLine(
+            p1,
+            p2,
+            shrinkFactor
+          );
+
+          newPointsData[outCount * 3] = shrunkP1[0];
+          newPointsData[outCount * 3 + 1] = shrunkP1[1];
+          newPointsData[outCount * 3 + 2] = shrunkP1[2];
+          const newId1 = outCount++;
+
+          newPointsData[outCount * 3] = shrunkP2[0];
+          newPointsData[outCount * 3 + 1] = shrunkP2[1];
+          newPointsData[outCount * 3 + 2] = shrunkP2[2];
+          const newId2 = outCount++;
+
+          // Add line
+          newLineData.push(2, newId1, newId2);
+        }
+        i += npts + 1;
+      }
+
+      newLines.setData(new Uint32Array(newLineData));
+    }
```

Comment:
looks like processing lines is the same code than processing polys, no ?
please factorize. (it can also be factorized with vertices)

Reply:
I’d prefer not to trade the code readability for a small reduction in duplication, since each type only needs a few lines.

Reply:
I do not see how it reduces code readability to create a function that "shrinks a cell" when the code is the same for 0D, 1D and 2D cells.
It makes it more readable because it shows that the process is the same for all cells.